File: app/src/main/java/com/example/mvvm_artichecture_sample/data/MainRepository.java
package com.example.mvvm_artichecture_sample.data;

import com.example.mvvm_artichecture_sample.base.BaseRepository;
import com.example.mvvm_artichecture_sample.base.network.RetrofitClient;
import com.example.mvvm_artichecture_sample.base.network.base.BaseRetrofitClient;
import com.example.mvvm_artichecture_sample.data.remote.ApiInterface;
import com.example.mvvm_artichecture_sample.data.remote.apimodel.CountryModel;

import io.reactivex.Observable;
import retrofit2.Response;

public class MainRepository extends BaseRepository implements MainRepositoryHandler {
    @Override
    protected BaseRetrofitClient getClient() {
        return null;
    }

    @Override
    public Observable<Response<CountryModel>> getList() {
        return RetrofitClient.getInstance().createRequest(ApiInterface.class).getList();
    }
}

File: app/src/main/java/com/example/mvvm_artichecture_sample/data/MainRepository.kt
package com.example.mvvm_artichecture_sample.data

import com.example.mvvm_artichecture_sample.base.BaseRepository
import com.example.mvvm_artichecture_sample.base.network.Output
import com.example.mvvm_artichecture_sample.base.network.RetrofitClient
import com.example.mvvm_artichecture_sample.data.remote.ApiInterface
import com.example.mvvm_artichecture_sample.data.remote.apimodel.ResponseModel

class MainRepository : BaseRepository<ApiInterface>(), MainRepositoryHandler {


    override suspend fun countries(requestType: String): Output<ResponseModel> {
        return safeApiCall(
                call = { client.getCountryAsync().await() },
                requestType = requestType
        )
    }

    override val client: ApiInterface
        get() = RetrofitClient.getInstance().createRequest(ApiInterface::class.java)

}
